File: packages/react-native/ReactCommon/react/coremodules/BatteryModule.cpp
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include "BatteryModule.h"
#include <ReactCommon/TurboModuleUtils.h>
#include <react/bridging/Bridging.h>

#ifdef _WIN32
#include <Windows.h>
#include <winrt/Windows.System.Power.h>
#include <winrt/base.h>
#include <stdexcept>

using namespace winrt;
using namespace winrt::Windows::System::Power;
#endif

namespace facebook::react {

BatteryState BatteryModule::getBatteryStateInternal() {
#ifdef _WIN32
  BatteryState state = {0, false, false};

  try {
    // Get battery report from Windows Power Manager
    auto batteryReport = PowerManager::BatteryReport();
    
    if (batteryReport) {
      auto fullChargeCapacity = batteryReport.FullChargeCapacityInMilliwattHours();
      auto remainingCapacity = batteryReport.RemainingCapacityInMilliwattHours();
      
      // Calculate battery level (0-100)
      if (fullChargeCapacity > 0) {
        state.level = static_cast<int>(
            (remainingCapacity * 100) / fullChargeCapacity
        );
        // Clamp to 0-100
        if (state.level < 0) state.level = 0;
        if (state.level > 100) state.level = 100;
      } else {
        // Fallback: use status to estimate
        auto status = batteryReport.Status();
        if (status == BatteryStatus::Charging || status == BatteryStatus::Idle) {
          state.level = 100; // Assume full when charging/idle
        } else {
          state.level = 50; // Unknown, use 50% as default
        }
      }

      // Determine charging status
      auto status = batteryReport.Status();
      state.isCharging = (status == BatteryStatus::Charging);

      // Check for energy saver (low power mode)
      state.isLowPowerMode = PowerManager::EnergySaverStatus() == 
                             EnergySaverStatus::On;
    } else {
      // No battery available (desktop PC, etc.)
      state.level = 100;
      state.isCharging = true; // Assume AC power
      state.isLowPowerMode = false;
    }
  } catch (const std::exception& e) {
    // Fallback values on error
    state.level = 100;
    state.isCharging = true;
    state.isLowPowerMode = false;
  }

  return state;
#else
  // Non-Windows platforms - return default values
  return {100, true, false};
#endif
}

jsi::Value BatteryModule::getBatteryState(jsi::Runtime &rt) {
  return createPromiseAsJSIValue(
      rt,
      [this](jsi::Runtime &runtime, std::shared_ptr<Promise> promise) {
        try {
          auto state = getBatteryStateInternal();
          promise->resolve(bridging::toJs(runtime, state, jsInvoker_));
        } catch (const std::exception& e) {
          promise->reject(e.what());
        }
      });
}

void BatteryModule::addListener(jsi::Runtime &rt, const std::string &eventName) {
  if (eventName != "batteryStateDidChange") {
    return;
  }

  if (!isObserving_) {
    isObserving_ = true;
    lastKnownState_ = getBatteryStateInternal();

#ifdef _WIN32
    try {
      // Listen for battery status changes
      batteryStatusChangedHandler_ = [this]() {
        auto newState = getBatteryStateInternal();
        
        // Only emit if state actually changed
        if (newState.level != lastKnownState_.level ||
            newState.isCharging != lastKnownState_.isCharging ||
            newState.isLowPowerMode != lastKnownState_.isLowPowerMode) {
          lastKnownState_ = newState;
          emitBatteryStateChange(newState);
        }
      };

      // Listen for energy saver status changes
      energySaverStatusChangedHandler_ = [this]() {
        auto newState = getBatteryStateInternal();
        lastKnownState_ = newState;
        emitBatteryStateChange(newState);
      };

      PowerManager::BatteryStatusChanged([this](auto&&, auto&&) {
        if (batteryStatusChangedHandler_) {
          batteryStatusChangedHandler_();
        }
      });

      PowerManager::EnergySaverStatusChanged([this](auto&&, auto&&) {
        if (energySaverStatusChangedHandler_) {
          energySaverStatusChangedHandler_();
        }
      });

      // Emit initial state
      emitBatteryStateChange(lastKnownState_);
    } catch (const std::exception& e) {
      // If event registration fails, still mark as observing
      isObserving_ = true;
    }
#else
    // Non-Windows: just emit initial state
    emitBatteryStateChange(lastKnownState_);
#endif
  }
}

void BatteryModule::removeListeners(jsi::Runtime &rt, double count) {
  if (count == 0) {
    isObserving_ = false;
#ifdef _WIN32
    batteryStatusChangedHandler_ = nullptr;
    energySaverStatusChangedHandler_ = nullptr;
#endif
  }
}

void BatteryModule::emitBatteryStateChange(BatteryState state) {
  if (!isObserving_) {
    return;
  }

  // Emit event to JavaScript using TurboModule event system
  emitDeviceEvent(
      "batteryStateDidChange",
      [state, jsInvoker = jsInvoker_](jsi::Runtime& rt, std::vector<jsi::Value>& args) {
        args.emplace_back(bridging::toJs(rt, state, jsInvoker));
      });
}

} // namespace facebook::react

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/modules/battery/BatteryModule.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.modules.battery

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.BatteryManager
import android.os.Build
import android.os.PowerManager
import com.facebook.fbreact.specs.NativeBatterySpec
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.WritableMap
import com.facebook.react.module.annotations.ReactModule

@ReactModule(name = NativeBatterySpec.NAME)
internal class BatteryModule(reactContext: ReactApplicationContext) :
    NativeBatterySpec(reactContext) {

  private var batteryReceiver: BroadcastReceiver? = null
  private var lastEmittedLevel: Int = -1
  private var lastEmittedCharging: Boolean = false
  private var lastEmittedLowPowerMode: Boolean = false

  override fun getBatteryState(promise: Promise) {
    try {
      val batteryState = getBatteryStateInternal()
      promise.resolve(batteryState)
    } catch (e: Exception) {
      promise.reject("BATTERY_ERROR", "Unable to get battery status", e)
    }
  }

  private fun getBatteryStateInternal(): WritableMap {
    val ifilter = IntentFilter(Intent.ACTION_BATTERY_CHANGED)
    val batteryStatus: Intent? = reactApplicationContext.registerReceiver(null, ifilter)

    if (batteryStatus == null) {
      throw Exception("Unable to get battery status")
    }

    val level = batteryStatus.getIntExtra(BatteryManager.EXTRA_LEVEL, -1)
    val scale = batteryStatus.getIntExtra(BatteryManager.EXTRA_SCALE, -1)
    val batteryPct = if (level >= 0 && scale > 0) {
      ((level / scale.toFloat()) * 100).toInt()
    } else {
      -1
    }

    val status = batteryStatus.getIntExtra(BatteryManager.EXTRA_STATUS, -1)
    val isCharging =
        status == BatteryManager.BATTERY_STATUS_CHARGING ||
        status == BatteryManager.BATTERY_STATUS_FULL

    val pm = reactApplicationContext.getSystemService(Context.POWER_SERVICE) as PowerManager
    val isLowPowerMode = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
      pm.isPowerSaveMode
    } else {
      false
    }

    val map = Arguments.createMap()
    map.putInt("level", batteryPct)
    map.putBoolean("isCharging", isCharging)
    map.putBoolean("isLowPowerMode", isLowPowerMode)

    return map
  }

  override fun addListener(eventName: String) {
    if (batteryReceiver == null) {
      batteryReceiver =
          object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
              if (Intent.ACTION_BATTERY_CHANGED == intent.action) {
                val batteryState = getBatteryStateInternal()
                val level = batteryState.getInt("level")
                val isCharging = batteryState.getBoolean("isCharging")
                val isLowPowerMode = batteryState.getBoolean("isLowPowerMode")

                // Throttle events: only emit if values actually changed
                if (lastEmittedLevel != level ||
                    lastEmittedCharging != isCharging ||
                    lastEmittedLowPowerMode != isLowPowerMode) {
                  lastEmittedLevel = level
                  lastEmittedCharging = isCharging
                  lastEmittedLowPowerMode = isLowPowerMode

                  val context = reactApplicationContext
                  if (context.hasActiveReactInstance()) {
                    context.emitDeviceEvent("batteryStateDidChange", batteryState)
                  }
                }
              }
            }
          }

      val filter = IntentFilter()
      filter.addAction(Intent.ACTION_BATTERY_CHANGED)
      // Also listen for power save mode changes on Android 5.0+
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
        filter.addAction(PowerManager.ACTION_POWER_SAVE_MODE_CHANGED)
      }
      reactApplicationContext.registerReceiver(batteryReceiver, filter)

      // Emit initial state
      val initialState = getBatteryStateInternal()
      val context = reactApplicationContext
      if (context.hasActiveReactInstance()) {
        context.emitDeviceEvent("batteryStateDidChange", initialState)
      }
    }
  }

  override fun removeListeners(count: Double) {
    if (batteryReceiver != null) {
      try {
        reactApplicationContext.unregisterReceiver(batteryReceiver)
      } catch (e: IllegalArgumentException) {
        // Receiver was not registered, ignore
      }
      batteryReceiver = null
      lastEmittedLevel = -1
      lastEmittedCharging = false
      lastEmittedLowPowerMode = false
    }
  }

  override fun invalidate() {
    super.invalidate()
    removeListeners(0.0)
  }

}

